### PR TITLE
feat: Light culling

### PR DESCRIPTION
**Pull request by MacXxs**
This PR implements light culling into the engine, optimizing lighting so it does not take a toll to performance and only renders visible lights.

|Without light culling|With light culling|
|----|----|
|image|image|

In the shown image there is an increase in performance of 30 fps (enabling ALL lights in the scene)

### CODE REVIEW

File: Source/DataModels/Render/LightPass.h [hunk 1]
```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include "GL/glew.h"
+
+#include "FileSystem/UID.h"
+
+#include "Resources/ResourceMesh.h"
+
+#include <map>
```
Comment (PaablooCH) [inline Source/DataModels/Render/LightPass.h:9]: this is inside the stdAfx, is necessary?

File: Source/Shaders/metallic_fragment.glsl
Comment (PaablooCH) [inline Source/Shaders/metallic_fragment.glsl:1]: removing light calculation from this fragment, and the specular one also, makes that the transparent objects don't receive light (transparents are not in gBuffer textures)

|Develop|this PR|
|----|----|
| ezgif com-video-to-gif (1)  | ezgif com-video-to-gif (2)

**Review by PaablooCH — CHANGES_REQUESTED** [review]
Found 1 error and other minor things

File: Source/Shaders/metallic_fragment.glsl
Comment (MacXxs) [reply Source/Shaders/metallic_fragment.glsl:1]: What we spoke in discord

File: Source/DataModels/Components/ComponentAreaLight.cpp [hunk 2]
```diff
@@ -9,6 +9,8 @@
 #include "Application.h"
 
 #include "Modules/ModuleScene.h"
+#include "Modules/ModuleDebugDraw.h"
```
[the remark this answers is not in the document]
Comment (MacXxs) [reply Source/DataModels/Components/ComponentAreaLight.cpp:12]: Removed it from the ifndef 'cause i need it in engine mode

File: Source/DataModels/Components/ComponentAreaLight.cpp [hunk 3]
```diff
@@ -169,21 +279,112 @@ void ComponentAreaLight::Draw() const
 	case AreaType::DISK:
 		break;
 	}
+
+#ifdef ENGINE
+	if (transform->IsDrawBoundingBoxes())
```
[the remark this answers is not in the document]
Comment (MacXxs) [reply Source/DataModels/Components/ComponentAreaLight.cpp:284]: 'Cause I only want to paint the bboxes in engine mode

File: Source/DataModels/Components/ComponentPointLight.cpp [hunk 4]
```diff
@@ -6,6 +6,8 @@
 #include "FileSystem/Json.h"
 
 #include "Modules/ModuleScene.h"
+#include "Modules/ModuleDebugDraw.h"
```
[the remark this answers is not in the document]
Comment (MacXxs) [reply Source/DataModels/Components/ComponentPointLight.cpp:9]: As stated before

File: Source/DataModels/Components/ComponentSpotLight.cpp [hunk 5]
```diff
@@ -6,6 +6,8 @@
 #include "FileSystem/Json.h"
 
 #include "Modules/ModuleScene.h"
+#include "Modules/ModuleDebugDraw.h"
```
[the remark this answers is not in the document]
Comment (MacXxs) [reply Source/DataModels/Components/ComponentSpotLight.cpp:9]: same

File: Source/DataModels/Render/LightPass.h (hunk 1 above)
Comment (MacXxs) [reply Source/DataModels/Render/LightPass.h:9]: We cannot include it in the .h, it should only be included in .cpp, if you test it you will see it crashes.

File: Source/DataModels/Components/ComponentAreaLight.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (PaablooCH) [reply Source/DataModels/Components/ComponentAreaLight.cpp:283]: my bad, I thought it was inside the destructor

**Review by AlexisDieu — APPROVED** [review]

**Review by PaablooCH — APPROVED** [review]

**Pull request merged**